favorites.py

```diff
@@ -272,29 +294,33 @@ def exists(self, s, group=False, group_name=None):
             # See if group exists
             return True if s in self.obj.files["groups"] else False
         else:
-            # See if file in global or group list exists
+            # See if file in global or group list exists, and return its index
             if group_name is None:
-                return True if s in set(self.obj.files["files"]) else False
+                i = [i for i, f in enumerate(self.obj.files["files"]) if f["file"] == s]
+                return i[0] if i else False
             else:
-                return True if s in set(self.obj.files["groups"][group_name]) else False
+                i = [i for i, f in enumerate(self.obj.files["groups"][group_name]) if f["file"] == s]
+                return i[0] if i else False
 

```

Comment:
changed to return the index because it's not a single string anymore, this is only used by remove because it needs the index

Reply:
I don't really care for this approach.  If it returns the index, then it should be called `get_index` or something like that. Also, if it is returning the index, it should return `None` if there is no index.  Also make sure that anyone who calls it is checking for `index is not None` and not for it's "truthiness" as an index of `0` as `0 == False` and is not `0 is False`.

If we are changing the the definition of what a function is doing, the name probably should change as well.

Reply:
Ok for everything else(renaming file to filename, using version 2, sanitize json).

This is a bit more problematic, for now I separated the function for group and file check, since exists() is also an import from os and they can be mistaken for each other, I renamed them in group_exists() (returns True or False) and file_index(returns index or None).

> Also make sure that anyone who calls it is checking for index is not None and not for its "truthiness" as an index of 0 as 0 == False and is not 0 is False.

I was already checking `if index is not False` where needed(line 310), now changing to `if index is not None`. But while the returned index is an integer and can be zero, it is returned if `i` is not null, and `i` itself is a list of indexes(line 302), so it wasn't causing problems even if it only contained [0]. `i` is only null if the list is empty, that is if the file isn't there. Also checked in the interpreter that [0] doesn't return False. Anyway I'm separating functions to avoid misinterpretations, if you agree.

Reply:
Okay, I took a closer look.  There should be a function if `group_exists` and `get_index` or something like that. I don't like the function conditionally returning different things. This will be confusing for people coming behind to make changes.

They should return either a boolean, or an index, but not both.  If it needs to be two separate functions, than that is fine.  I didn't realize that it checks if group *or* file exists (I was just looking at the diff).  These should now be two separate functions.

Also, I much prefer if it is returning a index to either return something like `-1` or `None` instead of `False`.

Reply:
I'm changing the function to:

```
    def exists(self, s, group=False, group_name=None):
        """Check if group or file exists."""

        if group:
            # See if group exists
            return True if s in self.obj.files["groups"] else False

        else:

            def is_present_in(key):
                for index, filename in enumerate(key):
                    if key['file'] == s:
                        return index
                return None

            # See if file in global or group list exists, and return its index
            if group_name is None:
                index = is_present_in(self.obj.files["files"])
            else:
                index = is_present_in(self.obj.files["groups"][group_name])

            return index if index is not None else None
```

So, no different names, just a bit clearer. Ok?

Reply:
But the function is  doing two different things now.  That's the problem.  In one case it is checking if it exists, in the second case it is actually returning/getting an index.

It's fine to have exists still work the same, you could even leverage the new `get_index` function.

```py
    def exists(self, s, group=False, group_name=None):
        """Check if group or file exists."""

        if group:
            # See if group exists
            return True if s in self.obj.files["groups"] else False

        else:

            return self.get_index(s, group, group_name) is not None
```

But I don't like it doing two different things.  You should call `self.get_index` when you want an index, and `self.exists` if you just want to know if it exists.  This will make it easier to for new contributors to understand in the future.

Reply:
I'm changing the function to:

```
    def group_exists(self, s):
        """Check if group exists."""

        return True if s in self.obj.files["groups"] else False

    def file_index(self, s, group_name=None):
        """Check if file exists, and return its index in this case."""

        def is_present_in(key):
            for index, filename in enumerate(key):
                if filename['file'] == s:
                    return index
            return None

        if group_name is None:
            index = is_present_in(self.obj.files["files"])
        else:
            index = is_present_in(self.obj.files["groups"][group_name])

        return index if index is not None else None
```

Reply:
Technically you can just return index as index is already either an index or None.

Also the parameter in `is_present_in` is technically not a key, but the value of the dictionary.

I know I did it this way `return True if s in self.obj.files["groups"] else False` but we can move to just `return s in self.obj.files["groups"]`.  The result is already a boolean. The whole if else isn't really needed. I was still learning Python and did unnecessary things here and there when I wrote this.

But yeah, generally, this looks good 🙂 .

Reply:
The linter might gripe that `is_present_in` is missing a docstring.  It's not really a public function, but the linter will gripe anyways.  Just an FYI.

Reply:
> Technically you can just return index as index is already either an index or None

I don't think so because if the index is zero it must be returned anyway in this case. If is_present_in() returns index zero, it must be returned.

> Also the parameter in is_present_in is technically not a key, but the value of the dictionary

Ok then value

> we can move to just return s in self.obj.files["groups"]

Ok. I'm still learning stuff myself, I'm not that advanced.

> The linter might gripe that is_present_in is missing a docstring

Will add.

Reply:
> I don't think so 

Maybe I wasn't clear what I was referring to:

```
return index if index is not None else None
```

If index is not None you are returning index.  If index is None, you are returning None.

So why not `return index`?  This is what I'm pointing out.  This line is redundant. Just return index.

Reply:
Yes I understood now, you're right. But maybe it's better like this, I can avoid checking `if index is not None` in the calling function, because zero would be also a valid value.

Reply:
I don't understand. Index == 0 is valid, I get that.

But:

```py
>>> index = 0
>>> index if index is not None
0
>>> index
0
>>> index = None
>>> index if index is not None
None
>>> index
None
```

You still get 0 - x, and `None` if it's `None`.  And you still have to check  check if index is `None` in the calling function. You aren't gaining anything by keeping it in `file_index`. If you don't want it to return `None`, the only option is to raise an exception and catch the exception or return an invalid index like `-1`.  But you still have to check if value is `>=0`.

Reply:
Yes sorry. Totally right, the check is still needed. Sorry again.

Reply:
No worries, like I said, there are probably still places in the code where I do this exact thing because I was learning 🙂 .

Reply:
I could push these changes, and I also added a check to remember project mode across session. I can include it now or leave it for later, what should I do?

Reply:
Let's handle persistent per project stuff separately.